// @xct007/frieren-scraper/dist/cjs/anime/doujindesu.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.detail = exports.search = exports.latest = void 0;
const https_1 = __importDefault(require("https"));
const Utils_1 = require("../Utils");
const Constant_1 = require("../Constant");
async function latest() {
    try {
        const { data } = await Utils_1.Axios.get(Constant_1.DoujindesuBaseUrl, {
            httpsAgent: new https_1.default.Agent({ keepAlive: true }),
            headers: {
                accept: "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7",
                "accept-language": "en-US,en;q=0.9,id;q=0.8,en-GB;q=0.7,en-GB-oxendict;q=0.6",
                "cache-control": "no-cache",
                pragma: "no-cache",
                "sec-ch-ua": '"Microsoft Edge";v="123", "Not:A-Brand";v="8", "Chromium";v="123"',
                "sec-ch-ua-mobile": "?0",
                "sec-ch-ua-platform": '"Windows"',
                "sec-fetch-dest": "document",
                "sec-fetch-mode": "navigate",
                "sec-fetch-site": "none",
                "sec-fetch-user": "?1",
                "sec-gpc": "1",
                "upgrade-insecure-requests": "1",
            },
        }).catch((e) => e?.response);
        const $ = (0, Utils_1.Cheerio)(data);
        const _temp = [];
        $(".feed#archives > .entries > .entry").each((i, e) => {
            const title = $(e).find("a").attr("title");
            const url = Constant_1.DoujindesuBaseUrl + $(e).find("a").attr("href");
            const thumbnail = $(e).find("img").attr("src");
            const chapter = $(e).find(".artists > a > span").text();
            _temp.push({
                title,
                chapter,
                thumbnail,
                url,
            });
        });
        if (Array.isArray(_temp) && _temp.length) {
            return _temp;
        }
        else {
            throw new Error("_temp is not an Array");
        }
    }
    catch (e) {
        return { error: true, message: e?.TypeError || String(e) };
    }
}
exports.latest = latest;
async function search(query) {
    try {
        const { data } = await Utils_1.Axios.get(Constant_1.DoujindesuBaseUrl, {
            params: {
                s: query,
            },
            httpsAgent: new https_1.default.Agent({ keepAlive: true }),
            headers: {
                accept: "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7",
                "accept-language": "en-US,en;q=0.9,id;q=0.8,en-GB;q=0.7,en-GB-oxendict;q=0.6",
                "cache-control": "no-cache",
                pragma: "no-cache",
                "sec-ch-ua": '"Microsoft Edge";v="123", "Not:A-Brand";v="8", "Chromium";v="123"',
                "sec-ch-ua-mobile": "?0",
                "sec-ch-ua-platform": '"Windows"',
                "sec-fetch-dest": "document",
                "sec-fetch-mode": "navigate",
                "sec-fetch-site": "none",
                "sec-fetch-user": "?1",
                "sec-gpc": "1",
                "upgrade-insecure-requests": "1",
            },
        }).catch((e) => e?.response);
        const $ = (0, Utils_1.Cheerio)(data);
        const _temp = [];
        $(".entries > .entry").each((i, e) => {
            const title = $(e).find("a").attr("title");
            const thumbnail = $(e).find("img").attr("src");
            const type = $(e).find(".type").text().trim();
            const score = $(e).find(".score").text().trim();
            const status = $(e).find(".status").text().trim();
            const url = Constant_1.DoujindesuBaseUrl + $(e).find("a").attr("href");
            _temp.push({ title, thumbnail, type, score, status, url });
        });
        if (Array.isArray(_temp) && _temp.length) {
            return _temp;
        }
        else {
            throw new Error("_temp is not an Array");
        }
    }
    catch (e) {
        return { error: true, message: e?.TypeError || String(e) };
    }
}
exports.search = search;
async function detail(url) {
    if (!url.includes(Constant_1.DoujindesuBaseUrl)) {
        return {
            error: true,
            message: "Invalid url!",
        };
    }
    try {
        const { data } = await Utils_1.Axios.get(url, {
            httpsAgent: new https_1.default.Agent({ keepAlive: true }),
            headers: {
                accept: "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7",
                "accept-language": "en-US,en;q=0.9,id;q=0.8,en-GB;q=0.7,en-GB-oxendict;q=0.6",
                "cache-control": "no-cache",
                pragma: "no-cache",
                "sec-ch-ua": '"Microsoft Edge";v="123", "Not:A-Brand";v="8", "Chromium";v="123"',
                "sec-ch-ua-mobile": "?0",
                "sec-ch-ua-platform": '"Windows"',
                "sec-fetch-dest": "document",
                "sec-fetch-mode": "navigate",
                "sec-fetch-site": "none",
                "sec-fetch-user": "?1",
                "sec-gpc": "1",
                "upgrade-insecure-requests": "1",
            },
        }).catch((e) => e?.response);
        const $ = (0, Utils_1.Cheerio)(data);
        const title = $(".thumbnail > a > img").attr("title").trim();
        const titles = $("h1 > span > i").text().trim();
        const thumbnail = $(".thumbnail > a > img").attr("src");
        const _metaData = [];
        $(".metadata > table > tbody > tr").each((i, e) => {
            const rowD = $(e).find("td");
            const rowA = [];
            rowD.each((_i, _e) => {
                rowA.push($(_e).text());
            });
            _metaData.push(rowA);
        });
        const metadata = _metaData.reduce((obj, [key, value]) => {
            obj[key.toLowerCase().replace(/\s/g, "_")] = value.trim();
            return obj;
        }, {});
        const _tags = [];
        $(".tags > a").each((i, e) => {
            _tags.push($(e).text().trim());
        });
        const tags = _tags.join(", ");
        const links = [];
        $(".linkdl > a").each((i, e) => {
            links.push({ title: $(e).attr("title"), url: $(e).attr("href") });
        });
        return {
            title,
            titles,
            tags,
            thumbnail,
            metadata,
            links,
        };
    }
    catch (e) {
        return { error: true, message: e?.TypeError || String(e) };
    }
}
exports.detail = detail;
